lib/copy/copy.js

```diff
@@ -109,18 +109,33 @@ function copyFileFallback (srcStat, src, dest, opts, cb) {
     const ws = fs.createWriteStream(dest, { mode: srcStat.mode })
     ws.on('error', err => cb(err))
       .on('open', () => rs.pipe(ws))
-      .once('close', () => setDestModeAndTimestamps(srcStat, dest, opts, cb))
+      .once('close', () => setDestModeAndTimestamps(srcStat, src, dest, opts, cb))
   })
 }
 
-function setDestModeAndTimestamps (srcStat, dest, opts, cb) {
-  fs.chmod(dest, srcStat.mode, err => {
+function setDestModeAndTimestamps (srcStat, src, dest, opts, cb) {
+  if (!opts.preserveTimestamps) {
+    return fs.chmod(dest, srcStat.mode, cb)
+  }
+
+  const next = (err) => {
     if (err) return cb(err)
-    if (opts.preserveTimestamps) {
-      return utimes(dest, srcStat.atime, srcStat.mtime, cb)
-    }
-    return cb()
-  })
+    // Cannot rely on previously fetched srcStat because atime is updated by
+    // the read operation: https://nodejs.org/docs/latest-v8.x/api/fs.html#fs_stat_time_values
+    return fs.stat(src, (err, srcStat) => {
+      if (err) return cb(err)
+      return utimes(dest, srcStat.atime, srcStat.mtime, (err) => {
+        if (err) return cb(err)
+        return fs.chmod(dest, srcStat.mode, cb)
+      })
+    })
+  }
+
+  if ((srcStat.mode & 0o200) === 0) {
+    // Make sure the file is writable first
+    return fs.chmod(dest, srcStat.mode | 0o200, next)
```

Comment:
This doesn't seem required! also this is a little hard to read. I think that'd be better we just run the functions without creating another inner function.

Reply:
See `copySync` discussion on the very same topic.

Reply:
Actually I think I managed to make this more readable and less confusing in f55125e. What do you think?